tests/zfs-tests/tests/functional/cache/cache_012_pos.ksh

```diff
@@ -0,0 +1,115 @@
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+# CDDL HEADER END
+#
+
+#
+# Copyright (c) 2020, George Amanakis. All rights reserved.
+#
+
+. $STF_SUITE/tests/functional/cache/cache.cfg
+. $STF_SUITE/tests/functional/cache/cache.kshlib
+
+#
+# DESCRIPTION:
+#	Looping around a cache device with l2arc_write_size exceeding
+#	the device size succeeds.
+#
+# STRATEGY:
+#	1. Create pool with a cache device.
+#	2. Set l2arc_write_max to a value larger than the cache device.
+#	3. Create a file larger than the cache device and random read
+#		for 10 sec.
+#	4. Verify that l2arc_write_max is set back to the default.
+#	5. Set l2arc_write_max to a value less than the cache device size but
+#		larger the default and closer to the size of the cache device.
+#		Subtract l2arc_write_boost to be safe in case the ARC has not
+#		warmed up yet.
+#	6. Record the l2_size.
+#	7. Random read for 15 sec to make sure we loop around without crashing.
+#	8. Record the l2_size again.
+#	9. If (6) <= (8) then we have not looped around yet.
+#	10. If (6) > (8) then we looped around. Break out of the loop and test.
+#	11. Destroy pool.
+#
+
+verify_runnable "global"
+
+log_assert "Looping around a cache device succeeds."
+
+function cleanup
+{
+	if poolexists $TESTPOOL ; then
+		destroy_pool $TESTPOOL
+	fi
+
+	log_must set_tunable32 L2ARC_WRITE_MAX $write_max
+	log_must set_tunable32 L2ARC_NOPREFETCH $noprefetch
+}
+log_onexit cleanup
+
+typeset write_max=$(get_tunable L2ARC_WRITE_MAX)
+typeset noprefetch=$(get_tunable L2ARC_NOPREFETCH)
+log_must set_tunable32 L2ARC_NOPREFETCH 0
+
+typeset fill_mb=$(( floor($MINVDEVSIZE * 3 / 4 ) ))
+export DIRECTORY=/$TESTPOOL
+export NUMJOBS=4
+export RUNTIME=10
+export PERF_RANDSEED=1234
+export PERF_COMPPERCENT=66
+export PERF_COMPCHUNK=0
+export BLOCKSIZE=128K
+export SYNC_TYPE=0
+export DIRECT=1
+export FILE_SIZE=$(( floor($fill_mb / $NUMJOBS) ))
+
+typeset VCACHE="$VDIR/vdev.cache"
+typeset VCACHE_SZ=$(( $MINVDEVSIZE / 2 ))
+
+log_must set_tunable32 L2ARC_WRITE_MAX $(( $VCACHE_SZ * 2 ))
+
+log_must truncate -s $VCACHE_SZ $VCACHE
+
+log_must zpool create $TESTPOOL $VDEV \
+	cache $VCACHE
+
+log_must fio $FIO_SCRIPTS/mkfiles.fio
+log_must fio $FIO_SCRIPTS/random_reads.fio
+
+typeset write_max2=$(get_tunable L2ARC_WRITE_MAX)
+
+log_must test $write_max2 -eq $write_max
+
+typeset write_boost=$(get_tunable L2ARC_WRITE_BOOST)
+typeset write_max2=$(( ($VCACHE_SZ * 3 / 4) - $write_boost ))
+log_must set_tunable32 L2ARC_WRITE_MAX $write_max2
+export RUNTIME=15
+
+typeset do_once=true
+while $do_once || [ $l2_size1 -le $l2_size2 ] ; do
+	typeset l2_size1=$(grep l2_size /proc/spl/kstat/zfs/arcstats \
+		| awk '{print $3}')
+	log_must echo $l2_size
+	log_must fio $FIO_SCRIPTS/random_reads.fio
+	typeset l2_size2=$(grep l2_size /proc/spl/kstat/zfs/arcstats \
```

Comment:
The /proc stuff only works on Linux. Can you hide this behind a function so it will be easier to adapt for other platforms?

Reply:
Sure!